Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1332 @@
+
+import Control.Monad ((<=<))
+import Language.Haskell.TH.Quote (QuasiQuoter (..))
+import Language.Haskell.TH.Syntax (Q, dataToPatQ, liftData)
+
+import SExp (SExp, parseSExp)
+
+sexp :: QuasiQuoter
+sexp = QuasiQuoter
+  { quoteExp  = liftData <=< parseOrThrow
+  , quotePat  = dataToPatQ (const Nothing) <=< parseOrThrow
+  , quoteType = const $ fail "No S-expression parser for type"
+  , quoteDec  = const $ fail "No S-expression parser for declarations"
+  }
+  where
+    parseOrThrow :: String -> Q SExp
+    parseOrThrow = either fail pure . parseSExp
+```
+
+Additionally, `liftData` and `dataToPatQ` assume that our parsed `SExp` implement the `Data` type class. We will change our `SExp` module to automatically derive this datatype for us. First, enable the `DeriveDataTypeable` language extension, so GHC can automatically implement `Data` for us.
+
+You also need to import `Data` from `Data.Data`.
+
+```hs
+{-# LANGUAGE DeriveDataTypeable #-}
+
+module SExp where
+
+import Data.Data (Data)
+-- The rest of the file goes here...
+```
+
+And now change `Identifier` and `SExp` to derive `Data`:
+
+```hs
+newtype Identifier = Identifier
+  { getId :: String
+  } deriving (Data, Show)
+
+data SExp
+  = SSExp    SExp [SExp]  -- (foo 0 "hello" bar), (bar (baz 1)), (foo)
+  | SInteger Integer  -- 42
+  | SString  String  -- "hello, world"
+  | SBool    Bool  -- false, true
+  | SId      Identifier  -- foo
+  | SDouble  Double  -- 42f, 3.1415f
+  deriving (Data, Show)
+```
+
+The basic contents for the `TH` module are summarized as follows:
+* Each field of `QuasiQuoter` provides a `String` parser for a different Haskell syntax.
+* `quoteExp` creates a Haskell expression.
+  * This will allow us to write expressions such as `let mySExp = [sexp|(foo 42)|]`.
+  * We use our standard Megaparsec parser, `parseSExp`, but fail if it can't parse (halting compilation). `liftData` then transforms our `SExp` into an `Exp`.
+  * If this is your first time seeing `<=<`, it's just another way to compose monadic functions. `g <=< f` is the same as `\x -> g =<< f x`.
+  * If you've read our previous TH tutorial, it creates an `Exp` just like `[e||]`.
+* `quotePat` creates a Haskell pattern.
+  * Example: `case mySExp of {[sexp|(foo 42)|] -> True; _ -> False}`.
+  * Similarly to `quoteExp`, we use `dataToPatQ (const Nothing)` to transform our `SExp` into a `Pat`.
+    * Confusingly, `liftData` is defined as `dataToExpQ (const Nothing)`.
+    * The `const Nothing` can be replaced to handle type-specific cases. We are fine with the default behavior for now but we'll change it later.
+  * Creates a `Pat`, like `[p||]`.
+* `quoteType` creates a Haskell type.
+  * Since we have no `SExp`, `SInteger` etc types matching our constructors, we will fail when someone tries to use our parser as types.
+  * Creates a `Type`, like `[t||]`.
+* `quoteDec` creates a Haskell declaration.
+  * Similarly to `quoteType`, we won't define this one.
+  * Creates a `Pat`, like `[p||]`.
```

Comment:
Copy-paste mistake?

Reply:
Good catch, it was a copy-paste mistake. I fixed it.